### ISSUE TITLE
Custom math operation over Tensor

### ISSUE THREAD

**Issue by janchk** [geohot/tinygrad#9336]
Hi. I'm experimenting with some quantization and would like to know how one could possibly implement some custom math operation for a tensor with custom gradient calculation/backward computation.

The way that I see right now is to:
1. Define some custom op in `tinygrad/ops.py`
2. Add entry to `tinygrad/gradient.py` to match that operation
3. Implement that operation for the child class of `Tensor`.

But I have no idea how the defined at step 1 op should match operation implementation at step 3.
Also I'm not a fan of modifying the framework's code, may be there is non-invasive approach that I didn't thought about.

**Comment by chenyuxyz** [comment]
we don't support custom now, what math operation do you want to add?

**Comment by janchk** [comment]
I want to add round operation with custom gradient calculation.

**Comment by janchk** [comment]
So I need myself a fork of tinygrad. Got it. But essentially I think I brought up some design/architecture-wise question.

**Comment by chenyuxyz** [comment]
if you have an example for how jax or torch gradient handles it, we are interested in matching that and update tinygrad. we are considering adding something like ceil/floor to ops though

generally we don't want to add more one-off op primitive if it can be represented with other ones, since we also need to implement it on each device

custom gradient is not supported now, but why would it be different from the auto grad one?

**Comment by janchk** [comment]
Well, talking about `round` operation, there is no correct way to calculate derivative of `y = round(x)` because it's obviously not differentiable. But there are some ways to estimate that derivative which implemented by the authors of some papers.

While doing some experiments in pytorch there is obvious ways to do it just by writing some logic in backward pass of nn.Module (like in 'pm_gradient' structure in tinygrad)

And when one building a computation graph with some dirty tricks in order to quantize network (for example) there are lot of custom logic involved.

**Comment by chenyuxyz** [comment]
interesting. yea i see torch is returning subgradient 0 everywhere, tinygrad is giving 1 now that i'd like to fix and match torch

we might consider custom api later, for now you can fork and add a custom op, and if the backward can be expressed with other ops in gradients.py, it should work

**Comment by janchk** [comment]
I got the idea.  Feel free to close the issue.
